src/FormParserFactory.php

```diff
@@ -0,0 +1,37 @@
+<?php
+
+namespace React\Http;
+
+use React\Http\Parser\FormUrlencoded;
+use React\Http\Parser\Multipart;
+use React\Http\Parser\NoBody;
+use React\Http\Parser\ParserInterface;
+
+class FormParserFactory
+{
+    /**
+     * @param Request $request
+     * @return ParserInterface
+     */
+    public static function create(Request $request)
+    {
+        $headers = $request->getHeaders();
+        $headers = array_change_key_case($headers, CASE_LOWER);
+
+        if (!array_key_exists('content-type', $headers)) {
+            return new NoBody($request);
+        }
+
+        $contentType = strtolower($headers['content-type']);
+
+        if (strpos($contentType, 'multipart/') === 0) {
+            return new Multipart($request);
+        }
+
+        if (strpos($contentType, 'application/x-www-form-urlencoded') === 0) {
+            return new FormUrlencoded($request);
+        }
+
+        return new NoBody($request);
```

Comment:
Ok maybe even better to not have much boilerplate code in user land:

Add in DeferredStream a third return value `body` which is filled when we can not decode the body. If `NoBody` all three values ('post', 'files', 'body') is empty. This way a user of this library can easily handle majority of requests. This way I don't have to handle stuff like `if (parser instanceof NoBody)` but just get three values when I want to have a application that is able to parse incoming requests.